81 stockpile fetch and deposit

Matte02 opened the pull request:
## Summary
Will make Pawns get materials to build and to empty their inventories while harvesting to the stockpile.

## Test Coverage
![image](https://user-images.githubusercontent.com/55716109/138562244-e99023e1-0409-4553-a869-00d7009e5600.png)
![image](https://user-images.githubusercontent.com/55716109/138562259-ac01870d-63b8-4ca9-98e4-13ec78014ae1.png)
![image](https://user-images.githubusercontent.com/55716109/138562268-45ebe5ff-b3d3-4531-b24e-b20072a14617.png)
![image](https://user-images.githubusercontent.com/55716109/138562273-ec02a553-8ad9-4d4d-a457-453759d8b270.png)
![image](https://user-images.githubusercontent.com/55716109/138562280-ea69cea5-72b5-4c21-9b5b-cc9c08e8b943.png)
![image](https://user-images.githubusercontent.com/55716109/138562285-979685ac-9d01-4a8d-a806-0c68d96db135.png)
![image](https://user-images.githubusercontent.com/55716109/138562297-d369872b-bcf7-4aa5-9fcb-db7bb7e20c67.png)

## Definition of Done
- [x] Completed all Tasks of the User Story
- [x] The Tasks have tests and they all pass
- [x] Overall test coverage is >= 90%
- [x] A UML class diagram of the newly implemented User Story is done and included in the README.md of the package

## Diagrams
Done, will be added when all other diagrams are updated.

Closes #81

Matte02 committed:
Add IReceiver

Matte02 committed:
Add IPositionable

Matte02 committed:
Make Stockpile implement IReceiver

Matte02 committed:
Make AbstractBeing implement IReceiver

Matte02 committed:
Add IGiver

Matte02 committed:
Make AbstractBeing and Stockpile implement IGiver

Matte02 committed:
Make IActionPerformer extend giver and receiver

Matte02 committed:
Add TakeItemAction

Matte02 committed:
Add GiveItemAction

Matte02 committed:
Add action in ActionFactory

Matte02 committed:
Add createMoveToStockpile

Matte02 committed:
Add move to stock and empty inven to HarvestRole

Matte02 committed:
Fix failing tests

Matte02 committed:
Add inventory tests for canItemFit

Matte02 committed:
Add structuretests for IGiver and IReceiver

Matte02 committed:
Add missing tests for Beings

Matte02 committed:
Extract getPosition method into IPositionable

Matte02 committed:
Add test for GiveItemAction

Matte02 committed:
Extract Inventory into Takeable Storeable

Matte02 committed:
Fix StockpileTests

Matte02 committed:
Remove IGiver and IReceiver

Matte02 committed:
Add stockpile tests

Matte02 committed:
Add BeingTests

Matte02 committed:
Rename tests in Give and Take Item action

Matte02 committed:
Fix add tests for HarvesterRoleTests

Matte02 committed:
Make pawns empty inventory at stockpile

Matte02 committed:
Add missing tests for inventory

Matte02 committed:
Fix last test in HarvesterRoleTests

Matte02 committed:
Fixed too long line in AbstractHarvesterRole

Matte02 committed:
Merge branch 'master' into 81-stockpile-fetch-and-deposit

Matte02 committed:
Change give and take item action to follow time

Matte02 committed:
Fix harvestRole to harvest and empty more than 1

Matte02 committed:
Add DoNextAction tests

Matte02 committed:
Add equals tests for give item action

Matte02 committed:
Change builder role to go to stockpile

Matte02 committed:
Add missing tests for actions

Matte02 committed:
Merge branch 'master' into 81-stockpile-fetch-and-deposit

Matte02 committed:
Fix leftovers from merge

Matte02 committed:
Make Town Hall a Stockpile

Matte02 committed:
Revert mistakenly modified codeStyleConfig

Matte02 committed:
Change naming from storeable to storable

Matte02 committed:
Change javadoc

Matte02 committed:
Fix bug with build role

Matte02 committed:
Revert accidental change to codeStyleConfig

JacobBredin commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -247,7 +245,10 @@ private void generateStructuresCache() {
+                              .filter(structure -> structure.getType().equals(type)
+                                                   || type.equals(StructureType.STOCKPILE)
+                                                      && structure.getType()
+                                                                  .equals(StructureType.TOWN_HALL))
```
Why is this needed? Or what is it for?

Matte02 replied:
It's used to allow the method to return Town Hall when searching for stockpiles.

JacobBredin replied:
Okay, now I see. 👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/IPositionable.java:
```diff
@@ -0,0 +1,15 @@
+package com.thebois.models;
+
+/**
+ * Something that has a position.
```
You can add author now that you created this

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+ * <p>
+ * It is an action given to a being when it asks for an action but has to skip the action and move
+ * on to next action.
+ * </p>
```
Author 🖊️

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+ *
+ * <p>
+ * Be careful when using this action as it may create an infinity loop. As this action can always be
+ * performed.
```
We say it can always be preformed, but `canPreform` returns false?

Matte02 replied:
Oh, should be changed to is always completed, good catch

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/GiveItemAction.java:
```diff
@@ -0,0 +1,64 @@
+import com.thebois.models.inventory.IStorable;
+
+/**
+ * Action give item from performer to a storable.
```
📖

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/GiveItemAction.java:
```diff
@@ -0,0 +1,64 @@
+public class GiveItemAction extends AbstractTimeAction implements Serializable {
+
+    private static final float TIME_REQUIRED_TO_GIVE_ITEM = 1f;
+    private static final float MINIMUM_GIVE_DISTANCE = 2f;
```
Seconds?
And 2 what?

Matte02 replied:
The number of tiles, or do we have a measurement for what 1 tile is?

martinjonsson01 commented on src/main/java/com/thebois/models/IPositionable.java:
```diff
@@ -0,0 +1,15 @@
+
+/**
+ * Something that has a position.
+ */
```
Missing author tag.

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/HarvestAction.java:
```diff
@@ -25,12 +25,15 @@ public HarvestAction(final IResource resource) {
     public boolean canPerform(final IActionPerformer performer) {
-        return resource.getPosition().distanceTo(performer.getPosition()) <= 1.0f;
+        final boolean isCloseEnough =
+            resource.getPosition().distanceTo(performer.getPosition()) <= 1.0f;
```
Question
Should we have a magic number here or is it not worth it?

Matte02 replied:
Should probably be changed like the other actions, menaing have a constant called mininum harvest distance or something. 
Changing it

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/actions/TakeItemAction.java:
```diff
@@ -0,0 +1,67 @@
+import com.thebois.models.inventory.items.ItemType;
+
+/**
+ * Action take takes an item from a takeable.
```
✏️

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -5,41 +5,57 @@
+    private boolean isEmptying = false;
 
     /**
      * Instantiates with a way of finding resources.
```
We have structures and resources now

martinjonsson01 commented on src/main/java/com/thebois/models/beings/IActionPerformer.java:
```diff
@@ -2,13 +2,15 @@
  * An internal interface that represents a performer of actions.
  */
-public interface IActionPerformer extends Serializable {
+public interface IActionPerformer extends Serializable, IStorable, ITakeable, IPositionable {
```
This, right here, is what we in the business call

***interface segregation***

👍 👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -62,10 +78,51 @@ private IAction createHarvestResource(final IActionPerformer performer) {
+        }
+        final Position position = performer.getPosition();
+        final Optional<IStructure> maybeStructure =
+            structureFinder.getNearbyStructureOfType(position, StructureType.STOCKPILE);
```
dosent change anything, but it removes a unneccesssary variable, we can just do 
` structureFinder.getNearbyStructureOfType(performer.getPosition(), StructureType.STOCKPILE);`
we dont seem to use it anywhere else in this method

Matte02 replied:
We are using it twice in the method and it makes the line less than 100 chars long, but can change it to `performer.getPosition()` if we want.

Matte02 replied:
Nvm it was in the method above, will change it in this one, and all others that only uses it once.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/ActionFactory.java:
```diff
@@ -67,4 +70,40 @@ public static IAction createBuild(final IStructure toBuild) {
+        return new TakeItemAction(takeable, itemType, position);
+    }
+
+    /**
+     * Creates an action of doing skipping an action do next in list.
+     *
+     * @return A do-next action.
+     */
+    public static IAction createDoNext() {
+        return new DoNextAction();
+    }
```
Ahh, that's how you solved it!

Nice, that's exactly what I was thinking `DoNothingAction` should've done in the beginning, but then someway along the line turned into an do-idle-stuff action instead.

Neat.

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -10,16 +10,23 @@
+    private static final int NUMBER_OF_BUILDING_STATES = 2;
     private final IStructureFinder finder;
     private final IWorld world;
+    private int buildingState = 0;
```
Feels a bit vague what these two mean, might be nice to have a comments explaining them 😄

Matte02 replied:
I tried giving them an explanation, I realize now I could use a boolean for the build state, needed an int before as I had 3 build state. Do we want to keep the int or change it to boolean?

martinjonsson01 replied:
Take a look at my suggestion for this below, before implementing it as a boolean.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/BuildAction.java:
```diff
@@ -52,20 +49,28 @@ public boolean equals(final Object other) {
+        final boolean performerHasNeededItem = getNeededItemFromPerformer(performer) != null;
+        final boolean isCloseEnough =
+            performer.getPosition().distanceTo(toBuild.getPosition()) < MINIMUM_BUILD_DISTANCE;
+        return isCloseEnough && performerHasNeededItem;
```
Supremely readable, nice job!

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+ * It is an action given to a being when it asks for an action but has to skip the action and move
+ * on to next action.
+ * </p>
+ */
```
Missing author tag.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+ * Makes the performer skip an action and move onto next action.
+ *
+ * <p>
+ * Be careful when using this action as it may create an infinity loop. As this action can always be
```
typo

infinity -> infinite

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+ *
+ * <p>
+ * Be careful when using this action as it may create an infinity loop. As this action can always be
+ * performed.
```
Actually, the javadoc is out of date. The implementation makes it so that the action can never be performed, but the javadoc says it can always be performed.

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+        final Position position = performer.getPosition();
+        final Optional<IStructure> maybeStructure =
+            finder.getNearbyStructureOfType(position, StructureType.STOCKPILE);
+        if (maybeStructure.isEmpty()) return ActionFactory.createDoNothing();
```
These are very similar across the a lot of methods, some parts seem pasted? Because once again we have position being used just once

Matte02 replied:
They are kind of copy pasted, I guess I could extract 
```
  final Optional<IStructure> maybeStructure =
            finder.getNearbyStructureOfType(position, StructureType.STOCKPILE);
        if (maybeStructure.isEmpty()) return ActionFactory.createDoNothing();
```

into a method and then check if the returned value is null keep going and if it is something else instantly return.  Didn't really like how that looked but maybe it is better than what we have now.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/DoNextAction.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
style

o -> other

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+        final Optional<IStructure> maybeStockpile =
+            finder.getNearbyStructureOfType(position, StructureType.STOCKPILE);
+        if (maybeStockpile.isEmpty()) return ActionFactory.createDoNothing();
+        final IStructure stockpile = maybeStockpile.get();
```
Exact same snippet of code in my last comment, only change is maybestockpile

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/FisherRole.java:
```diff
@@ -12,11 +13,13 @@
-     * @param finder The locator of resources.
-     * @param world  The world in which the resources are located.
+     * @param finder          The locator of resources.
+     * @param structureFinder The locator of structures.
```
since we call it `structureFinder` maybe `finder` should be `resourceFinder`, the same in the other files we added this

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/LumberjackRole.java:
```diff
@@ -12,11 +13,13 @@
      *
-     * @param finder The locator of resources.
-     * @param world  The world in which the resources are located.
+     * @param finder          The locator of resources.
```
resourceFinder

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/roles/MinerRole.java:
```diff
@@ -12,11 +13,15 @@
      *
-     * @param finder The locator of resources.
-     * @param world  The world in which the resources are located.
+     * @param finder          The locator of resources.
```
resourceFinder

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/GiveItemAction.java:
```diff
@@ -0,0 +1,64 @@
+     * @param storablePosition Where the storable is in the world.
+     */
+    public GiveItemAction(
+        final IStorable storable, final Position storablePosition) {
```
question

It's a bit confusing that the action is `GiveItem` but in the constructor you never specify which item to give. So is it more like `GiveAnyItem`? Or how does it pick which item to give?

Matte02 replied:
Yeah, right now it's any item, should probably swap it to so you have to give an item type to remove, just like the TakeItemAction.

martinjonsson01 replied:
Yeah, if you've got the time then that'd be better.

Would also eliminate the need for the `getNextItem` method, right?

Matte02 replied:
Depends if we want to loop through the inventory to get the next item to remove when emptying inventory for builders.

martinjonsson01 replied:
What's wrong with looping? Am I sensing premature optimization 👀

Matte02 replied:
`performer.takeFirstItem()` vs 
`for (ItemType itemType : ItemType.values()) {
            if (performer.hasItem(itemType)) {
                return itemType;
            }
}`

I guess it's not that bad now that I've written that out, but I feel like wouldn't be unreasonable to have a `takeFirstItem` on an `Inventory`, or am I wrong?

martinjonsson01 replied:
Isn't that for-loop exactly what `tryTake(ItemType)` does?

martinjonsson01 replied:
So since you're changing the action to only give items of a specific type, then it should be 
`performer.tryTake(itemType)` vs `performer.takeFirstItem()`.

Which means that `takeFirstItem` is redundant, because there's never a use case where a client doesn't care what item type it's taking.

Matte02 replied:
When the `BuilderRole` is emptying the inventory then it doesn't care what item's they are removing as they are removing all items, so inside of `BuilderRole` I'm just doing `takeFirstItem` and giving that `ItemType` to the `GiveItemAciton`, but if we want to loop inside `BuilderRole` then I can change it to that.

martinjonsson01 replied:
At the beginning of this conversation we discussed how `GiveItemAction` should only give *specific types* of items:

>Yeah, right now it's any item, should probably swap it to so you have to give an item type to remove, just like the TakeItemAction.

So what you're saying here:

>When the `BuilderRole` is emptying the inventory then it doesn't care what item's they are removing as they are removing all items

no longer applies, right? That's how it worked before. But now we want `GiveItemAction` to only give items of a specific type. So the `BuilderRole` *does* care, or have I misunderstood?

I'm talking about the `GiveItemAction`, are you talking about another action `GiveAnyItemAction`?

Matte02 replied:
Right now `GiveAnyItemAction` and `BuilderRole` is using the `GiveItemAction` and the item given to `GivenItemAction` is decided with the `takeFirstItem`.

Want me to add the `GiveAnyItemAction` and use that instead?

martinjonsson01 replied:
So you're saying there is a `GiveAnyItemAction` that delegates to `GiveItemAction`? But then you're asking if you should add `GiveAnyItemAction`, so it's not a thing? I'm completely lost.

Let's start over.

>question
It's a bit confusing that the action is GiveItem but in the constructor you never specify which item to give. So is it more like GiveAnyItem? Or how does it pick which item to give?

>Yeah, right now it's any item, should probably swap it to so you have to give an item type to remove, just like the TakeItemAction.

Are we or are we not making `GiveItemAction` take in an itemtype in its constructor? And if we are, then what is the need for `takeFirstItem`, since it's only used inside the `GiveItemAction` to get around the fact that it doesn't know which item to ask for.

martinjonsson01 replied:
Actually, this isn't really important enough to spend too much time discussing. Keep it as-is if you want, we should focus on merging this PR instead.

Matte02 replied:
`GiveItemAction` is now changed to take in `ItemType` in the constructor.

Then in `BuilderRole` in order to decide what `Item` to remove, it would be nice to have `getFirstTypeOfItem`(just realized I can't use `takeFirstItem` cause that removes same type of item twice without being added twice.) or something since when emptying the inventory the `BuilderRole` doesn't care what `Item` to remove just that all `Item`s should be removed.

If I'm not making myself understood, we could start a zoom meeting and discuss it there, if you have time.

jonathan-lindqvist commented on src/main/java/com/thebois/models/inventory/IInventory.java:
```diff
@@ -1,89 +1,10 @@
-     *
-     * @return The counter.
-     */
-    int numberOf(ItemType itemType);
```
🤩

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/TakeItemAction.java:
```diff
@@ -0,0 +1,67 @@
+
+/**
+ * Action take takes an item from a takeable.
+ */
```
Missing author tag.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/actions/TakeItemAction.java:
```diff
@@ -0,0 +1,67 @@
+     * @param takeablePosition The position of the takeable.
+     */
+    public TakeItemAction(
+        final ITakeable takeable, final ItemType itemType, final Position takeablePosition) {
```
I'm noticing that you're always passing `IStoreable` and `ITakeable` together with positions, is that always the case, or are there use-cases where you don't care about the position of the "storable"/"takable" and only want to access it? 

I'm imagining that you always want to know where you need to be to give/take something, so maybe these two interfaces should extends `IPositionable`, what do you think?

Matte02 replied:
It was more the case of inventory doesn't have a position, so takeable and storable should probably not have a position. I guess I could make an interface that extends takeable and positionable but I don't know about that.

martinjonsson01 replied:
Ah, that's true. It sounds a bit weird for an inventory to have a position...

Unless? 

Actually, is there a case where you want to access an inventory without knowing if you're near enough to be able to access it? Isn't it always the case that you also want to check the distance to the holder of the inventory?

Matte02 replied:
Hmm
Actually, you are probably right, but I feel like it would be a pain the ass to make sure the position of the inventory is always the same as the pawn. I Will take a look at this when I've handled the rest of the review.

martinjonsson01 replied:
Good call, this isn't necessary at all and should be very low priority.

Oh, yeah, the inventory still shouldn't have its own position. Since an inventory is always a part of some entity, the position it returns should be the position of the entity. 

But I see what you're saying, that still means that the inventory needs a reference to the entity, which is whack. Maybe `IPositionable` is what saves the day here, since the "holder" of the inventory (i.e. `Being` or `Structure`) always implements it, then the inventory can just take an `IPositionable` in its constructor, and whenever inventory is asked for its position, it just delegates to the injected positionable. 

But, maybe we don't need to make `IInventory` have a position at all. We could make a third interface `IStoreableLocation extends IStorable, IPositionable` which is what is taken in by the constructor above. In that case, `Stockpile` would have to implement `IStoreableLocation` instead, while `IInventory` still can have no position.

I don't know, maybe this is all just making it unnecessarily complicated. It just seemed a bit smelly to me that we had to pass around a position next to the interfaces all the time.

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/resources/Water.java:
```diff
@@ -13,7 +13,7 @@
      * The time it takes to harvest, in seconds.
      */
-    private static final float HARVEST_TIME = 3f;
+    private static final float HARVEST_TIME = 0.3f;
```
isnt this super fast?

Matte02 replied:
Not really if you take into consideration that the pawn can atm carry 200 fishes.

martinjonsson01 replied:
We can probably save the balance-tweaking until tomorrow when we go through it all before turning it in.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -5,41 +5,57 @@
     @Override
     protected Collection<IActionSource> getTaskGenerators() {
-        return List.of(this::createMoveToResource, this::createHarvestResource);
+        if (!isEmptying) {
```
style

You should switch around the if-statement so you don't need the negation. It's easier to read `if(isEmptying)` than to interpret it as `if (not isEmptying)`.

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/Stockpile.java:
```diff
@@ -1,16 +1,21 @@
  * A structure used for storing items.
  */
-class Stockpile extends AbstractStructure {
+class Stockpile extends AbstractStructure implements ITakeable, IStorable {
```
Shouldn't this just implement `IInventory ` since thats exactly all it does now?

Matte02 replied:
Very true, changing it

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -5,41 +5,57 @@
+        }
+        else {
+            return List.of(this::createMoveToStockpile, this::createEmptyInventoryOfResource);
+        }
```
This turned out really clean, nicely done!

jonathan-lindqvist reviewed (changes requested):
![bild](https://user-images.githubusercontent.com/47867973/138563899-019229da-6af8-4cfb-b963-3f013e7373aa.png)
Great work, just look over the comments, shouldn't be anything major! 😄

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -62,10 +78,51 @@ private IAction createHarvestResource(final IActionPerformer performer) {
+        if (closestSpotNextToStructure.isEmpty()) return ActionFactory.createDoNothing();
+
+        return ActionFactory.createMoveTo(closestSpotNextToStructure.get());
+    }
```
question

This looks like a copy/paste of `createMoveToResource`, were you not able to factor the common parts out into a single method? 

Ah, I suppose that's really difficult since they use different finders and different interfaces returned.

Matte02 replied:
That was indeed the problem.

martinjonsson01 replied:
The thing is, whenever you encounter a bunch of methods doing the same thing in the exact same order and same structure, but using minorly different variables or types, that is a sign that there is an underlying abstraction waiting to be created. 

If we could identify the common idea between all of these methods, then that would be the missing abstraction. Would make the roles a whole lot cleaner. 

But this isn't the worst thing ever, so leave it as-is. I'll try to come up with a nice hotfix if I can figure out the underlying abstraction, but we might not be able to fix this before the deadline.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/AbstractHarvesterRole.java:
```diff
@@ -62,10 +78,51 @@ private IAction createHarvestResource(final IActionPerformer performer) {
+            return ActionFactory.createDoNothing();
+        }
+
+        return ActionFactory.createGiveItem(storable, structure.getPosition());
```
style

Looks a bit weird to have the declaration of `storable` split from its assignment, just to have the return statement at the end of the method.

Could you move the return statement that uses the `storable` into the inside of the if-statement? Maybe java complains if you do that.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+            return List.of(this::createMoveToStockpile,
+                           this::createEmptyInventory,
+                           this::createFillInventory);
+        }
```
"buildingState" is quite vague, it's hard to understand what it really means for a builder to be in a specific "building state". Could you maybe use a private enum instead? Something like

```java
private enum BuilderState {
    FETCHING_RESOURCES = 0,
    BUILDING_STRUCTURE = 1
}
```

I think you should still be able to increment the enum with `buildingState++`, since enums really are 16-bit numbers with titles. 

It would make the code a lot more readable if it said `if (buildingState == BuilderState.FETCHING_RESOURCES)` rather than what it is right now.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+            return ActionFactory.createDoNothing();
+        }
+        return ActionFactory.createMoveTo(closestSpotNextToStructure.get());
+    }
```
Maybe I'm misremembering but don't you have the exact same method in `AbstractHarvesterRole`? If that is the case, then this should be broken out into a method in `ActionFactory` called `createMoveToStockpile` or something similar.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+        }
+
+        return ActionFactory.createGiveItem(storable, structure.getPosition());
+    }
```
Same here, isn't this used in two places in the exact same way?

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+            if (!performer.hasItem(type, structure.getNumberOfNeedItemType(type))) {
+                nextNeededItem = type;
+            }
+        }
```
style

Should be broken out into a method.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/roles/BuilderRole.java:
```diff
@@ -40,34 +47,138 @@ public RoleType getType() {
+            return ActionFactory.createDoNext();
+        }
+
+        return ActionFactory.createTakeItem(takeable, nextNeededItem, stockpile.getPosition());
```
Here you can make it a lot more readable as well by moving the parts on line 133-139 into the inside of the if-clause on line 127.

martinjonsson01 commented on src/main/java/com/thebois/models/inventory/IStorable.java:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * Methods for storing items.
+ */
```
Missing author tag.

martinjonsson01 commented on src/main/java/com/thebois/models/inventory/ITakeable.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * Methods for taking items and checking if they exists.
+ */
```
Missing author.

martinjonsson01 commented on src/main/java/com/thebois/models/inventory/ITakeable.java:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @return The item to take.
+     */
+    IItem takeNextItem();
```
style

I've seen you use this method quite a bit, but I always asked the question "what qualifies as the *next* item?", and here I finally see that it's the *first* item in the inventory.

Maybe the method should be called `takeFirstItem` or `takeTopItem`.

martinjonsson01 commented on src/main/java/com/thebois/models/world/ITile.java:
```diff
@@ -2,19 +2,12 @@
 /**
  * A section of the world representing a structure or resource.
  */
-public interface ITile extends Serializable {
-
-    /**
-     * Gets the current location.
-     *
-     * @return The current location.
-     */
-    Position getPosition();
```
One more step closer to perfection!

martinjonsson01 commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -247,7 +245,10 @@ private void generateStructuresCache() {
-                              .filter(structure -> structure.getType().equals(type))
+                              .filter(structure -> structure.getType().equals(type)
+                                                   || type.equals(StructureType.STOCKPILE)
+                                                      && structure.getType()
```
style

This whole filter-conditional should be broken out into a method with a clear name like `isStorable(structure)`.

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -64,8 +64,8 @@ private int countAllNeededItems() {
-            final int numberOfNeededItems = allNeededItems.get(itemType) - deliveredItems.numberOf(
-                itemType);
+            final int numberOfNeededItems =
+                allNeededItems.get(itemType) - deliveredItems.numberOf(itemType);
```
![image](https://user-images.githubusercontent.com/12572888/138564616-b187cb6a-fe86-4f22-808a-75d3dd4e3061.png)

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -74,6 +74,11 @@ private int countAllNeededItems() {
     }
 
+    public int getNumberOfNeedItemType(final ItemType itemType) {
+        if (!allNeededItems.containsKey(itemType)) return 0;
```
Is this just in `AbstractStructure` and not in `IBuildable`? If that is the case, then it should probably be added to `IBuildable` and an `@Override` tag should be here, if it already is, then it's missing an override tag.

Matte02 replied:
Yup, was just missing the `@Override` tag

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/IBuildable.java:
```diff
@@ -18,6 +18,15 @@
     Collection<ItemType> getNeededItems();
 
+    /**
+     * Counts the amount of times still needed to build the structure.
```
style

Not really the amount of times, rather the amount of items.

Matte02 replied:
Let's just say that items and times are almost spelled the same. And me being tired missed that I misspelled it.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -42,6 +42,7 @@
     private static final float HUNGER_RATE = 1f;
     private AbstractRole role;
     private AbstractBeing being;
+    private static final IInventory inventory = mock(IInventory.class);
```
It's a bit iffy to have a static mock, because then it's the same for all the different tests. 

It's better to have it be re-instantiated in a setup-method with `@BeforeEach` so that a new inventory is used for each test.

Matte02 replied:
I'll take a look at this.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -487,4 +505,137 @@ public AbstractRole deepClone() {
+
+        // Assert
+        verify(inventory, times(1)).isEmpty();
+    }
```
The best kind of tests 😩 

Hey, at least we can actually verify that it delegates, that's something!

martinjonsson01 commented on src/test/java/com/thebois/models/beings/actions/GiveItemActionTests.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Test
+    public void onPerformRemovesItemFromPerformerAndAddsItemToStorable() {
+        // Arrange
```
style

No need for empty clauses of the triple-A test structure.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/actions/TakeItemActionTests.java:
```diff
@@ -0,0 +1,179 @@
+
+    @Test
+    public void performTakesItemFromTakeableAndGivesToPerformer() {
+        // Arrange
+        when(takeable.hasItem(any())).thenReturn(true);
+        when(takeable.take(any())).thenReturn(ItemFactory.fromType(itemType));
+        when(performer.canFitItem(any())).thenReturn(true);
+        when(performer.getPosition()).thenReturn(position);
+
+        // Act
+        action.perform(performer, 5f);
+
+        // Assert
+        verify(takeable, times(1)).take(any());
+        verify(performer, times(1)).tryAdd(any());
```
Not gonna lie, I really like this exact way of structuring tests.

1) The clear AAA-structure.
2) The clutter of instantiation moved to a setup-method.
3) The use of clear variable names.
4) Mocks with very clear purpose.
5) Clear assertions in the form of verify-calls.

Great job.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/roles/BuilderRoleTests.java:
```diff
@@ -55,129 +66,483 @@ public void teardown() {
+    }
+
+    @Test
+    public void obtainNextActionReturnsIdleActionWhenOneRotationHaveGoneAndNoStockpileCouldBeFoundToFill() {
```
Really great test names, all around. I'm able to tell what you're testing in each method without having to actually read the method, and if I am intrigued by the name then I can read the test for an exact description of how its done. 

Well done. 
![image](https://user-images.githubusercontent.com/12572888/138565076-83360d79-b434-4000-b5d1-985be91f818e.png)

martinjonsson01 reviewed (approved):
Phenomenal work, this was a real complex story but you got it to function while still testing everything! 

Left a bunch of suggestions, but I think it'd be okay even if we had to merge it as-is.

![image](https://user-images.githubusercontent.com/12572888/138565224-3b223890-38ed-47ad-b851-d733532502b9.png)

Matte02 committed:
Add author tag and fix javadoc/comments

Matte02 committed:
Extract methods

Matte02 committed:
Clean up code to be less messy

Matte02 committed:
Add enum for builderState

Matte02 committed:
Make inventory not static in BeingTests

Matte02 committed:
Change give GiveItemAction to require an ItemType

Matte02 committed:
Extract more actions from role to ActionFactory

Matte02 committed:
Fix small bugs

Matte02 committed:
Add javadoc for GiveItemAction variable

Matte02 committed:
Add some more missing javadoc

Matte02 committed:
Add some author tags

Matte02 committed:
Fix bug in GiveItemAction

Matte02 committed:
Fix bug for builders

The bug was that builders where empyting and filling the inventory at the same making the builder stuck in a loop.

Matte02 committed:
Merge branch 'master' into 81-stockpile-fetch-and-deposit

Matte02 committed:
Add 1 more state to builderRole

Matte02 committed:
Make harveserts empty whole inventory

This was changed inorder to not force the player to change harvesters to builders in order for them
to emtpy inventory to be able to gather another resource.

Matte02 committed:
Fix bug for builder going to unfinished stockpile

martinjonsson01 commented on src/main/java/com/thebois/models/inventory/ITakeable.java:
```diff
@@ -72,10 +72,10 @@
 
     /**
-     * Gets the top most item.
+     * Gets the ItemType of the first item.
```
style

of *any* item

martinjonsson01 reviewed (approved):
Looking good!

Sat and actually played the game for a good 10 minutes and it's actually quite entertaining now!

Didn't find any bugs other than the ones Jacob already mentioned.

Matte02 committed:
Move IPositionable to abstractions

Matte02 committed:
Change javadoc

Matte02 committed:
Merge branch 'master' into 81-stockpile-fetch-and-deposit

Matte02 committed:
Fix Stockpile bug and too long lines

Matte02 committed:
Add missing javadoc from last change

martinjonsson01 commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -247,16 +247,19 @@ private void generateStructuresCache() {
                                                              origin.distanceTo(o2.getPosition())));
     }
 
+    private boolean isSameType(final StructureType type, final IStructure structure) {
```
This name doesn't reflect the fact that it also checks for `|| isStorable(structure, type)`

jonathan-lindqvist reviewed (approved):
LGTM!
![image](https://user-images.githubusercontent.com/47867973/138600538-7fc20fa3-6236-4f53-ab63-06c83cd01d7a.png)

Matte02 committed:
Change method name
